File: cv/nms/nms.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits>

#define eps (std::numeric_limits<float>::epsilon())
struct Box
{
	/**
	 *  o--------> x
	 *  |
	 *  |
	 *  |y
	 */
	float x_tl;
	float y_tl;
	float x_br;
	float y_br;
	float conf_score;

	Box(float _x_tl, float _y_tl, float _x_br, float _y_br, float _conf_score)
		: x_tl(_x_tl), y_tl(_y_tl), x_br(_x_br), y_br(_y_br)
	{}
};


float iou(const Box& box_one, const Box& box_two)
{
	float x_tl_i = std::max<float>(box_one.x_tl, box_two.x_tl);
	float y_tl_i = std::max<float>(box_one.y_tl, box_two.y_tl);

	float x_br_i = std::min<float>(box_one.x_br, box_two.x_br);
	float y_br_i = std::min<float>(box_one.y_br, box_two.y_br);

	float iou = 0.0f;
	float x_len_i = std::abs(x_br_i - x_tl_i);
	float y_len_i = std::abs(y_br_i - y_tl_i);

	/* if (std::abs(x_br_i - x_tl_i) > eps && std::abs(y_br_i - y_tl_i) > eps) */
	if (x_len_i > eps && y_len_i > eps)
	{
		float inter = x_len_i * y_len_i;
		float uni = std::abs((box_one.x_br - box_one.x_tl) * (box_one.y_br - box_one.y_tl))
				+ std::abs((box_two.x_br - box_two.x_tl) * (box_two.y_br - box_two.y_tl)) - inter;

		iou = inter / (uni + eps);
	}

	return iou;
}

void nms(std::vector<Box>& boxes, const float& iou_thres, std::vector<Box>& keeped_boxes)
{
	/* sort the box in non-increasing order of conf_score */
	std::sort(boxes.begin(), boxes.end(),
			[](const Box& lhs, const Box& rhs)
			{
				// larger score is placed in front.
				return lhs.conf_score > rhs.conf_score;
			});

	/* Compute IOU. */
	int box_cnt = boxes.size();
	std::vector<std::vector<float>> iou_mat(box_cnt, std::vector<float>(box_cnt, 0.0f));
	for (int r = 0; r < box_cnt; ++r)
	{
		for (int c = r + 1; c < box_cnt; ++r)
		{
			iou_mat[r][c] = iou(boxes[r], boxes[c]);
		}
	}

	std::vector<bool> is_excluded(boxes.size(), false);

	/* Iterate the box in increasing order of confidence score. */
	for (int i = 0; i < box_cnt; ++i)
	{
		if (is_excluded[i]) continue;

		/* Check all boxes with samller score*/
		for (int j = i + 1; j < box_cnt; ++j)
		{
			if (is_excluded[j]) continue;

			if (iou_mat[i][j] > iou_thres)
			{
				is_excluded[j] = true;
			}
		}
	}

	keeped_boxes.clear();
	for (int i = 0; i < box_cnt; ++i)
	{
		if (!is_excluded[i])
		{
			keeped_boxes.emplace_back(std::move(boxes[i]));
		}
	}
}

int main()
{
	return 0;
}

File: cv/nms/nms.py
# class Box:
#     def __init__(self, x_tl : float, y_tl : float,
#                  x_br : float, y_br : float, score : float) -> None:
#         self.x_tl = x_tl
#         self.y_tl = y_tl
#         self.x_br = x_br
#         self.y_br = y_br
#         self.score = score

import numpy as np
from functools import cmp_to_key

def get_iou(box_one : np.ndarray, box_two : np.ndarray):
    x_inter_tl = max(box_one[0], box_two[0])
    y_inter_tl = max(box_one[1], box_two[1])

    x_inter_br = min(box_one[2], box_two[2])
    y_inter_br = min(box_one[3], box_two[3])

    inter_area = max(0.0, x_inter_br - x_inter_tl) * max(0.0, y_inter_br - y_inter_tl)
    one_area = (box_one[2] - box_one[0]) * (box_one[3] - box_one[1])
    two_area = (box_two[2] - box_two[0]) * (box_two[3] - box_two[1])

    eps = 1e-9
    iou = inter_area / (one_area + two_area - inter_area + eps)

    return iou

def nms(boxes: list, iou_thres: float):
    # boxes is of shape (N, 5)
    # boxes[4] is the score

    # https://numpy.org/doc/stable/reference/generated/numpy.sort.html
    # sort along axis means sorting across that axis
    # sorte a MxN array, axis = 0 means accros rows -> sort each column.
    # while axis = 1 means crossing columns -> sort each row

    # boxes[:, 4] is of shape (N,) and sort in increasing order and then reverse
    # sorted_idx = np.argsort(boxes[:, 4])[::-1]

    sorted_idx = sorted(range(len(boxes)), key=lambda i : boxes[i][4], reverse=True)

    picked_boxes = []
    while len(sorted_idx) > 0:
        best_idx = sorted_idx[0]
        picked_boxes.append(boxes[best_idx])

        sorted_idx = [i for i in sorted_idx[1:] if get_iou(boxes[best_idx, i] <= iou_thres)]

    return picked_boxes


if __name__ == '__main__':
    pass
